hierarchicalforecast/methods.py

```diff
@@ -988,6 +1005,116 @@ def fit_predict(self,
 
     __call__ = fit_predict
 
+@njit(parallel=True, fastmath=True, error_model="numpy")
+def _shrunk_covariance_schaferstrimmer_no_nans(Xs: np.ndarray, mint_shr_ridge: float):
+    """Shrink empirical covariance according to the following method:
+        Schäfer, Juliane, and Korbinian Strimmer. 
+        ‘A Shrinkage Approach to Large-Scale Covariance Matrix Estimation and 
+        Implications for Functional Genomics’. Statistical Applications in 
+        Genetics and Molecular Biology 4, no. 1 (14 January 2005). 
+        https://doi.org/10.2202/1544-6115.1175.
+
+    :meta private:
+    """
+    n_timeseries = Xs.shape[0]
+    n_samples = Xs.shape[1]
+    
+    # We need the empirical covariance, the off-diagonal sum of the variance of 
+    # the empirical correlation matrix and the off-diagonal sum of the squared 
+    # empirical correlation matrix.
+    W = np.zeros((n_timeseries, n_timeseries), dtype=np.float64).T
+    sum_var_emp_corr = np.float64(0.0)
+    sum_sq_emp_corr = np.float64(0.0)
+    factor_emp_corr = np.float64(n_samples / (n_samples - 1))
+    factor_shrinkage = np.float64(1 / (n_samples * (n_samples - 1)))
+    epsilon = np.float64(2e-8)
+    for i in prange(n_timeseries):
+        # Mean of the standardized residuals
+        Xs_i_mean = np.mean(Xs[i])
+        for j in range(i + 1):
+            # Empirical covariance
+            W[i, j] = factor_emp_corr * np.mean(Xs[i] * Xs[j])
```

Comment:
I think we may be inadvertently allocating $W$ as the correlation matrix here as this is the covariance of the standardised residuals, right? I thought ``Xs`` are the standardised deviations from the means. It might be worth checking if we are actually shrinking the correlation matrix instead of the covariance matrix by accident.

Reply:
Yeah, I think I'm mixing it, I already had this feeling after one of the revisions I did. I need to see how to correct it, but good obvervation

Reply:
Solved it. This also explains the performance delta between the versions I was observing. Now the legacy version and new implementation give nearly identical results when tested on an actual dataset (synthetic was equal already).